server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java

```diff
@@ -298,21 +336,94 @@ public void updateTaskResult(ReportTaskRunModel ce) {
      */
-    public WfRunIdModel getWfRunId() {
-        return taskRunSource.getSubSource().getWfRunId();
+    public void markAttemptReadyToSchedule() {
+        dispatchTaskToQueue();
+    }
+
+    /**
+     * Dispatches the latest TaskAttempt to the server's Task Queues.
+     */
+    public void dispatchTaskToQueue() {
+        if (attempts.isEmpty()) {
+            throw new IllegalStateException("can't have empty attempts queue");
+        }
+        TaskAttemptModel attempt = attempts.get(attempts.size() - 1);
+        if (attempt.getStatus() != TaskStatus.TASK_PENDING) {
+            throw new IllegalStateException("Cannot schedule task attempt that isn't in PENDING state");
+        }
+
+        attempt.setStatus(TaskStatus.TASK_SCHEDULED);
+        attempt.setScheduleTime(new Date());
+
+        ScheduledTaskModel scheduledTask = new ScheduledTaskModel();
+        scheduledTask.setVariables(inputVariables);
+        scheduledTask.setAttemptNumber(attempts.size() - 1);
+        scheduledTask.setCreatedAt(new Date());
+        scheduledTask.setSource(taskRunSource);
+        scheduledTask.setTaskDefId(taskDefId);
+        scheduledTask.setTaskRunId(id);
+
+        processorContext.getTaskManager().scheduleTask(scheduledTask);
+    }
+
+    private boolean shouldRetry() {
+        // Shouldn't look at nodeRun to decide whether to retry task or not.
+        TaskAttemptModel latest = getLatestAttempt();
+        if (latest == null) {
+            // This really shouldn't happen I think
+            return false;
+        }
+
+        if (latest.getStatus() != TaskStatus.TASK_FAILED && latest.getStatus() != TaskStatus.TASK_TIMEOUT) {
+            // Can only retry timeout or task failure.
+            return false;
+        }
+
+        switch (retryPolicyType) {
+            case SIMPLE_TOTAL_ATTEMPTS:
+                return simpleTotalAttempts > attempts.size();
+            case EXPONENTIAL_BACKOFF:
+                return exponentialBackoffRetryPolicy
+                        .calculateDelayForNextAttempt(attempts.size())
+                        .isPresent();
+            case RETRYPOLICY_NOT_SET:
+        }
+        return false;
+    }
+
+    private void scheduleRetryAtAppropriateTime() {
+        if (retryPolicyType == RetryPolicyCase.SIMPLE_TOTAL_ATTEMPTS) {
+            attempts.add(new TaskAttemptModel());
+            dispatchTaskToQueue();
+            return;
+        }
+
+        if (retryPolicyType != RetryPolicyCase.EXPONENTIAL_BACKOFF) {
+            throw new IllegalStateException("Unimplemented retry policy type");
+        }
+
+        long delayMs = exponentialBackoffRetryPolicy
+                .calculateDelayForNextAttempt(attempts.size())
+                .get();
+        Date maturationTime = new Date(System.currentTimeMillis() + delayMs);
+        LHTimer timer = new LHTimer(new CommandModel(new TaskAttemptRetryReadyModel(id), maturationTime));
+        processorContext.getTaskManager().scheduleTimer(timer);
```

Comment:
`public void scheduleTimer(CoreSubcommand commandTimer, delayMs) {}`
This method in the task manager class could be useful: